File: go/core/m2m_auth_provider_test.go
// Copyright (c) 2026 Omas Cloud
//
// SPDX-License-Identifier: MIT

package core

import (
	"context"
	"errors"
	"io"
	"net/http"
	"strings"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

type roundTripperFunc func(*http.Request) (*http.Response, error)

func (f roundTripperFunc) RoundTrip(request *http.Request) (*http.Response, error) {
	return f(request)
}

type mutableClock struct {
	mu  sync.Mutex
	now time.Time
}

func (clock *mutableClock) Now() time.Time {
	clock.mu.Lock()
	defer clock.mu.Unlock()
	return clock.now
}

func (clock *mutableClock) Advance(duration time.Duration) {
	clock.mu.Lock()
	defer clock.mu.Unlock()
	clock.now = clock.now.Add(duration)
}

func tokenResponse(token string, expiresIn int) *http.Response {
	return &http.Response{
		StatusCode: http.StatusOK,
		Header:     http.Header{},
		Body: io.NopCloser(strings.NewReader(
			`{"accessToken":"` + token + `","tokenType":"Bearer","expiresIn":` + formatValue(expiresIn) + `}`)),
	}
}

func TestNewM2MAuthProviderValidatesOptions(t *testing.T) {
	if _, err := NewM2MAuthProvider(" "); err == nil {
		t.Fatal("blank credential accepted")
	}
	if _, err := NewM2MAuthProvider("credential", WithM2MRefreshSkew(-time.Second)); err == nil {
		t.Fatal("negative refresh skew accepted")
	}
	if _, err := NewM2MAuthProvider("credential", WithM2MHTTPClient(nil)); err == nil {
		t.Fatal("nil HTTP client accepted")
	}
}

func TestM2MAuthProviderExchangesAndCachesTokenByService(t *testing.T) {
	clock := &mutableClock{now: time.Date(2026, 8, 2, 12, 0, 0, 0, time.UTC)}
	var calls atomic.Int32
	httpClient := &http.Client{Transport: roundTripperFunc(func(request *http.Request) (*http.Response, error) {
		calls.Add(1)
		body, err := io.ReadAll(request.Body)
		if err != nil {
			t.Fatal(err)
		}
		if request.Method != http.MethodPost || request.URL.String() != tokenExchangeEndpoint ||
			string(body) != `{"audience":"metrics"}` {
			t.Fatalf("unexpected exchange request: %s %s %s", request.Method, request.URL, body)
		}
		if request.Header.Get("Authorization") != "Bearer credential" ||
			request.Header.Get("Content-Type") != "application/json" {
			t.Fatalf("unexpected exchange headers: %v", request.Header)
		}
		return tokenResponse("access-token", 300), nil
	})}

	provider, err := NewM2MAuthProvider(
		" credential ",
		WithM2MHTTPClient(httpClient),
		withM2MClock(clock.Now),
	)
	if err != nil {
		t.Fatal(err)
	}
	for range 2 {
		authentication, resolveErr := provider.Resolve(context.Background(), AuthContext{Service: "metrics"})
		if resolveErr != nil {
			t.Fatal(resolveErr)
		}
		if authentication.Headers.Get("Authorization") != "Bearer access-token" {
			t.Fatalf("authorization = %s", authentication.Headers.Get("Authorization"))
		}
	}
	if calls.Load() != 1 {
		t.Fatalf("exchange calls = %d", calls.Load())
	}
	if _, err := provider.Resolve(context.Background(), AuthContext{Service: " "}); err == nil {
		t.Fatal("blank service accepted")
	}
}

func TestM2MAuthProviderCoalescesConcurrentRefresh(t *testing.T) {
	var calls atomic.Int32
	started := make(chan struct{})
	release := make(chan struct{})
	httpClient := &http.Client{Transport: roundTripperFunc(func(*http.Request) (*http.Response, error) {
		if calls.Add(1) == 1 {
			close(started)
		}
		<-release
		return tokenResponse("shared-token", 300), nil
	})}
	provider, err := NewM2MAuthProvider("credential", WithM2MHTTPClient(httpClient))
	if err != nil {
		t.Fatal(err)
	}

	const callers = 20
	errorsByCaller := make(chan error, callers)
	for range callers {
		go func() {
			_, resolveErr := provider.Resolve(context.Background(), AuthContext{Service: "metrics"})
			errorsByCaller <- resolveErr
		}()
	}
	<-started
	close(release)
	for range callers {
		if err := <-errorsByCaller; err != nil {
			t.Fatal(err)
		}
	}
	if calls.Load() != 1 {
		t.Fatalf("exchange calls = %d", calls.Load())
	}
}

func TestM2MAuthProviderRefreshWaiterCanCancel(t *testing.T) {
	started := make(chan struct{})
	release := make(chan struct{})
	httpClient := &http.Client{Transport: roundTripperFunc(func(*http.Request) (*http.Response, error) {
		close(started)
		<-release
		return tokenResponse("shared-token", 300), nil
	})}
	provider, err := NewM2MAuthProvider("credential", WithM2MHTTPClient(httpClient))
	if err != nil {
		t.Fatal(err)
	}
	leaderDone := make(chan error, 1)
	go func() {
		_, resolveErr := provider.Resolve(context.Background(), AuthContext{Service: "metrics"})
		leaderDone <- resolveErr
	}()
	<-started

	waiterContext, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
	defer cancel()
	_, err = provider.Resolve(waiterContext, AuthContext{Service: "metrics"})
	if !errors.Is(err, context.DeadlineExceeded) {
		t.Fatalf("waiter error = %v", err)
	}
	close(release)
	if err := <-leaderDone; err != nil {
		t.Fatal(err)
	}
}

func TestM2MAuthProviderUsesValidTokenAfterEarlyRefreshFailure(t *testing.T) {
	clock := &mutableClock{now: time.Date(2026, 8, 2, 12, 0, 0, 0, time.UTC)}
	var calls atomic.Int32
	httpClient := &http.Client{Transport: roundTripperFunc(func(*http.Request) (*http.Response, error) {
		if calls.Add(1) == 1 {
			return tokenResponse("initial-token", 100), nil
		}
		return nil, errors.New("exchange unavailable")
	})}
	provider, err := NewM2MAuthProvider(
		"credential",
		WithM2MHTTPClient(httpClient),
		WithM2MRefreshSkew(30*time.Second),
		withM2MClock(clock.Now),
	)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := provider.Resolve(context.Background(), AuthContext{Service: "metrics"}); err != nil {
		t.Fatal(err)
	}
	clock.Advance(75 * time.Second)
	authentication, err := provider.Resolve(context.Background(), AuthContext{Service: "metrics"})
	if err != nil {
		t.Fatal(err)
	}
	if authentication.Headers.Get("Authorization") != "Bearer initial-token" {
		t.Fatalf("authorization = %s", authentication.Headers.Get("Authorization"))
	}

	clock.Advance(26 * time.Second)
	_, err = provider.Resolve(context.Background(), AuthContext{Service: "metrics"})
	var authenticationError *AuthenticationError
	if !errors.As(err, &authenticationError) {
		t.Fatalf("expired token error = %T %v", err, err)
	}
}

func TestM2MAuthProviderRejectsInvalidExchangeResponseWithoutLeakingCredential(t *testing.T) {
	httpClient := &http.Client{Transport: roundTripperFunc(func(*http.Request) (*http.Response, error) {
		return &http.Response{
			StatusCode: http.StatusOK,
			Header:     http.Header{},
			Body:       io.NopCloser(strings.NewReader(`{"accessToken":"credential-secret","tokenType":"Basic","expiresIn":0}`)),
		}, nil
	})}
	provider, err := NewM2MAuthProvider("credential-secret", WithM2MHTTPClient(httpClient))
	if err != nil {
		t.Fatal(err)
	}
	_, err = provider.Resolve(context.Background(), AuthContext{Service: "metrics"})
	if err == nil || strings.Contains(err.Error(), "credential-secret") {
		t.Fatalf("unsafe error = %v", err)
	}
}

type closeAwareTransport struct{ closed atomic.Int32 }

func (*closeAwareTransport) RoundTrip(*http.Request) (*http.Response, error) {
	return tokenResponse("token", 300), nil
}
func (transport *closeAwareTransport) CloseIdleConnections() { transport.closed.Add(1) }

func TestM2MAuthProviderDoesNotCloseInjectedClient(t *testing.T) {
	transport := &closeAwareTransport{}
	provider, err := NewM2MAuthProvider(
		"credential",
		WithM2MHTTPClient(&http.Client{Transport: transport}),
	)
	if err != nil {
		t.Fatal(err)
	}
	if err := provider.Close(); err != nil {
		t.Fatal(err)
	}
	if transport.closed.Load() != 0 {
		t.Fatalf("injected transport closed %d times", transport.closed.Load())
	}
}

File: typescript/core/src/m2m-auth-provider.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";

import { AuthenticationError } from "./errors.js";
import { M2mAuthProvider } from "./m2m-auth-provider.js";

const context = { service: "metrics", operationId: "listMetrics" };

describe("M2mAuthProvider", () => {
    afterEach(() => {
        vi.useRealTimers();
    });

    it.each([
        "",
        "   ",
        "\n\t",
    ])("rejects blank credential %#", (credential) => {
        expect(() => new M2mAuthProvider(credential)).toThrow(
            "credential must not be blank",
        );
    });

    it.each([
        -1,
        Number.NaN,
        Number.POSITIVE_INFINITY,
    ])("rejects invalid refresh skew %s", (refreshSkewMs) => {
        expect(
            () => new M2mAuthProvider("credential", { refreshSkewMs }),
        ).toThrow("refreshSkewMs must be a non-negative finite number");
    });

    it("rejects a blank target service without exchanging", async () => {
        let exchangeCalls = 0;
        const provider = new M2mAuthProvider("credential", {
            fetch: async () => {
                exchangeCalls += 1;
                return new Response(null, { status: 500 });
            },
        });

        await expect(
            provider.resolve(
                { service: "  ", operationId: "listMetrics" },
                new AbortController().signal,
            ),
        ).rejects.toThrow("auth context service must not be blank");
        expect(exchangeCalls).toBe(0);
    });

    it("exchanges once and caches a service-scoped bearer token", async () => {
        vi.useFakeTimers();
        vi.setSystemTime(new Date("2026-08-03T00:00:00Z"));
        const requests: Array<{ url: string; init?: RequestInit }> = [];
        const provider = new M2mAuthProvider(" workspace-credential ", {
            fetch: async (input, init) => {
                requests.push({ url: String(input), init });
                return new Response(
                    JSON.stringify({
                        accessToken: "access-token",
                        tokenType: "Bearer",
                        expiresIn: 300,
                    }),
                    {
                        status: 200,
                        headers: { "Content-Type": "application/json" },
                    },
                );
            },
        });
        const signal = new AbortController().signal;

        const first = await provider.resolve(context, signal);
        const second = await provider.resolve(context, signal);

        expect(new Headers(first.headers).get("Authorization")).toBe(
            "Bearer access-token",
        );
        expect(new Headers(second.headers).get("Authorization")).toBe(
            "Bearer access-token",
        );
        expect(requests).toHaveLength(1);
        expect(requests[0]?.url).toBe(
            "https://api.omas.cloud/v1/auth/token-exchange",
        );
        expect(requests[0]?.init?.method).toBe("POST");
        expect(
            new Headers(requests[0]?.init?.headers).get("Authorization"),
        ).toBe("Bearer workspace-credential");
        expect(requests[0]?.init?.body).toBe('{"audience":"metrics"}');
    });

    it("keeps independent token caches for different services", async () => {
        const audiences: string[] = [];
        const provider = new M2mAuthProvider("credential", {
            fetch: async (_input, init) => {
                const body = JSON.parse(String(init?.body)) as {
                    audience: string;
                };
                audiences.push(body.audience);
                return new Response(
                    JSON.stringify({
                        accessToken: `${body.audience}-token`,
                        tokenType: "Bearer",
                        expiresIn: 300,
                    }),
                    { status: 200 },
                );
            },
        });
        const signal = new AbortController().signal;

        await provider.resolve(context, signal);
        await provider.resolve(
            { service: "operations", operationId: "getOperation" },
            signal,
        );
        await provider.resolve(context, signal);

        expect(audiences).toEqual(["metrics", "operations"]);
    });

    it("coalesces concurrent refreshes for one service", async () => {
        let exchangeCalls = 0;
        let releaseExchange: (() => void) | undefined;
        const exchangeStarted = new Promise<void>((resolve) => {
            releaseExchange = resolve;
        });
        const provider = new M2mAuthProvider("credential", {
            fetch: async () => {
                exchangeCalls += 1;
                await exchangeStarted;
                return new Response(
                    JSON.stringify({
                        accessToken: "shared-token",
                        tokenType: "Bearer",
                        expiresIn: 300,
                    }),
                    { status: 200 },
                );
            },
        });
        const signal = new AbortController().signal;

        const resolutions = Array.from({ length: 20 }, () =>
            provider.resolve(context, signal),
        );
        await vi.waitFor(() => expect(exchangeCalls).toBe(1));
        releaseExchange?.();

        const authentications = await Promise.all(resolutions);
        expect(exchangeCalls).toBe(1);
        expect(
            authentications.map((authentication) =>
                new Headers(authentication.headers).get("Authorization"),
            ),
        ).toEqual(Array.from({ length: 20 }, () => "Bearer shared-token"));
    });

    it("retains a valid token after failed early refresh but never after expiry", async () => {
        vi.useFakeTimers();
        const startedAt = new Date("2026-08-03T00:00:00Z");
        vi.setSystemTime(startedAt);
        let exchangeCalls = 0;
        const provider = new M2mAuthProvider("credential", {
            fetch: async () => {
                exchangeCalls += 1;
                if (exchangeCalls > 1) {
                    throw new Error("exchange unavailable");
                }
                return new Response(
                    JSON.stringify({
                        accessToken: "original-token",
                        tokenType: "Bearer",
                        expiresIn: 100,
                    }),
                    { status: 200 },
                );
            },
        });
        const signal = new AbortController().signal;

        await provider.resolve(context, signal);
        vi.setSystemTime(new Date(startedAt.getTime() + 70_000));
        const duringEarlyRefreshFailure = await provider.resolve(
            context,
            signal,
        );

        expect(
            new Headers(duringEarlyRefreshFailure.headers).get("Authorization"),
        ).toBe("Bearer original-token");
        vi.setSystemTime(new Date(startedAt.getTime() + 100_000));
        await expect(provider.resolve(context, signal)).rejects.toMatchObject({
            constructor: AuthenticationError,
            message: "authenticate listMetrics request",
        });
        expect(exchangeCalls).toBe(3);
    });

    it("allows a refresh waiter to cancel without cancelling the shared exchange", async () => {
        let releaseExchange: (() => void) | undefined;
        const exchangeGate = new Promise<void>((resolve) => {
            releaseExchange = resolve;
        });
        const provider = new M2mAuthProvider("credential", {
            fetch: async () => {
                await exchangeGate;
                return new Response(
                    JSON.stringify({
                        accessToken: "shared-token",
                        tokenType: "Bearer",
                        expiresIn: 300,
                    }),
                    { status: 200 },
                );
            },
        });
        const first = provider.resolve(context, new AbortController().signal);
        const waiterController = new AbortController();
        const waiter = provider.resolve(context, waiterController.signal);
        const reason = new Error("caller stopped waiting");

        waiterController.abort(reason);

        await expect(waiter).rejects.toMatchObject({
            constructor: AuthenticationError,
            cause: reason,
        });
        releaseExchange?.();
        await expect(first).resolves.toMatchObject({
            headers: { Authorization: "Bearer shared-token" },
        });
    });

    it.each([
        [{ accessToken: "", tokenType: "Bearer", expiresIn: 300 }],
        [{ accessToken: "token", tokenType: "bearer", expiresIn: 300 }],
        [{ accessToken: "token", tokenType: "Bearer", expiresIn: 0 }],
        [{ accessToken: "token", tokenType: "Bearer", expiresIn: 1.5 }],
    ])("rejects malformed token exchange payload %#", async (payload) => {
        const provider = new M2mAuthProvider("credential", {
            fetch: async () =>
                new Response(JSON.stringify(payload), {
                    status: 200,
                    headers: { "Content-Type": "application/json" },
                }),
        });

        await expect(
            provider.resolve(context, new AbortController().signal),
        ).rejects.toMatchObject({
            constructor: AuthenticationError,
            message: "authenticate listMetrics request",
        });
    });

    it("redacts the credential and response body from exchange errors", async () => {
        const credential = "workspace-super-secret";
        const returnedToken = "returned-super-secret";
        const provider = new M2mAuthProvider(credential, {
            fetch: async () =>
                new Response(
                    JSON.stringify({
                        error: `${credential} ${returnedToken}`,
                        errorCode: "ACCESS_DENIED",
                    }),
                    { status: 401 },
                ),
        });

        const rejected = provider.resolve(
            context,
            new AbortController().signal,
        );
        await expect(rejected).rejects.toBeInstanceOf(AuthenticationError);
        try {
            await rejected;
        } catch (error) {
            const serialized = String(error);
            expect(serialized).not.toContain(credential);
            expect(serialized).not.toContain(returnedToken);
        }
    });
});
